Add support for TAP 15

MVrachev opened the pull request:
Related to issue: #1909

**Description of the changes being introduced by the pull request**:

[TAP 15](https://github.com/theupdateframework/taps/blob/master/tap15.md) was created on June 23-rd 2020 and was last modified in July 6-th 2020.
Since then the TAP has been put to `draft` status meaning it needs a prototype implementation before it's accepted as a future specification change.
Given that this TAP underlines an efficient way of handling hash bin delegations, meaning it would be really useful when `python-tuf` is integrated into `Warehouse`, it's logical that we should not only create a prototype but directly work to integrate it in `python-tuf`.

The outcome of the TAP 15 implementation is
1. Support reading, updating, and saving target metadata files that are using `succinct hashed bin delegations`.
2. Support in ngclient for calculating the delegated role name on a given target path.
3. An exemplary document inside `examples/` showcasing how `succinct hashed bin delegations` can be utilized in practice.

This pr covers points 1 and 2.
**Point 3 will be done in another pr.** 

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
DelegatedRole: support succinct_hash_delegations

Start supporting succinct_hash_delegations attribute inside
DelegatedRole. This change is part of TAP 15 implementation.

Unfortunatly, this change is not compatible with all previos python-tuf
versions as the "name" attribute is described as optional in TAP 15,
meaning it has to be replaced to become the fourth argument
instead of being the first when creating a DelegatedRole instance.

Additionally, serialization tests covering the new use cases as TAP 15
states that only one of "paths", "path_hash_prefixes" and
"succinct_hash_delegations" is allowed in a DelegatedRole instance.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Tests delegation graphs: make paths an optional

Inside test_updater_delegation_graphs.py we define the TestDelegation
class. There the "paths" attribute is defined as a "List[str]" and
it has a default value of a "[*]". This is often used so that all
targets are considered as delegated by the delegated role that will
consequently be created with a paths attribute with a value of "[*]".
Still, there are a couple of test cases where we want to give a "None"
value to the "paths" attribute and mypy complains that we can't assign
a "None" value when the expected type is a list of strings.
That's why we have to make "paths" optional.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
ngclient: add support for TAP 15

As part of the implementation of TAP 15 I had to make sure that ngclient
supports target metadata files using succinct hash delegations.

Some things to keep in mind while reviewing:
- first, it should be noted that when succinct hash info is added inside
a particular delegated role this means that every possible target has
a bin assigned to it
- the "name" attribute for DelegatedRole objects has been changed to
optional inside TAP 15, because when succinct hash delegations is used
"name" becomes redundant
- I have added an additional helper function
`DelegatedRole._calculate_rolename" for easier testing
- I intentially added more tests inside the "get_rolename_dataset"
to showcase how succinct hash delegations is using binary prefix instead
of hash prefix.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
This pr is a draft because of warnings given by `mypy`.
Those warnings are connected with the issue discussed in https://github.com/theupdateframework/python-tuf/issues/1943.

I am not sure what is the problem for python3.7 https://github.com/theupdateframework/python-tuf/runs/5888251078?check_suite_focus=true

jku commented on tuf/api/metadata.py:
```diff
@@ -1274,12 +1274,14 @@ class DelegatedRole(Role):
     instance attributes.*
 
     Args:
-        name: Delegated role name.
```
this is a backwards incompatible API change, we might not want that

MVrachev replied:
Yea, you are correct. We should start thinking as a real library and be careful with such changes.
Because of Python limitations, I had to move the `name` argument if it has a default value after all attributes without a default value.
I think maybe it will be better to require a value for `name` even if it's `None` and not move it.
Will try with that.

jku commented on tuf/api/metadata.py:
```diff
@@ -1289,32 +1291,59 @@ class DelegatedRole(Role):
+        name: Optional[str] = None,
         paths: Optional[List[str]] = None,
         path_hash_prefixes: Optional[List[str]] = None,
+        succinct_hash_info: Optional[Dict[str, Any]] = None,
```
is there a reason for using a different name than the TAP does?

MVrachev replied:
Yes, it boils down to that it's just shorter.
`succinct_hash_delegations` is 25 characters and because of our characters per line restriction, I had to often split lines.
`succinct_hash_info` is 18 characters which save a good amount of characters while still hinting at the purpose of this argument.

jku commented on tuf/api/metadata.py:
```diff
@@ -1384,6 +1419,46 @@ def _is_target_in_pathpattern(targetpath: str, pathpattern: str) -> bool:
+            return self.name
+
+        # number of bins
+        bit_length = self.succinct_hash_info["delegation_hash_prefix_len"]
```
Dictionaries that pretend to be data structures is a code smell to me: it prevents static typing checks and makes typos deadly. There's multiple other ways to handle this and I think we should consider them:
* maybe hash_prefix_len and bin_name_prefix are just instance variables in DelegatedRole? 
* Maybe succinct_hash_delegations is a data class of it's own -- bit overkill but fits the Metadata API design?
* Maybe DelegatedRole is actually three separate classes for paths, prefixes and succinct prefixes cases? This also sounds like OO-overkill but I'll mention it anyway: all the classes would just implement `get_rolename()` and the de/serialization separately

jku commented on tuf/ngclient/updater.py:
```diff
@@ -438,11 +438,10 @@ def _preorder_depth_first_walk(
                 for child_role in targets.delegations.roles.values():
                     if child_role.is_delegated_path(target_filepath):
-                        logger.debug("Adding child role %s", child_role.name)
+                        child_name = child_role.get_rolename(target_filepath)
```
Is is_delegated_path() useful for anything now? It seems we only call it in combination with get_rolename(). 

So we could just have get_rolename() return `Optional[str]` and just have a single public method in DelegatedRole: no need to call is_delegated_path() anymore?

jku commented on tuf/api/metadata.py:
```diff
@@ -1384,6 +1419,46 @@ def _is_target_in_pathpattern(targetpath: str, pathpattern: str) -> bool:
+        return f"{name_prefix}-{bin_number}"
+
+    def get_rolename(self, target_filepath: str) -> str:
+        """Get the delegated rolename based on the given ``target filepath``.
```
the name is techincally correct but feels a bit off (we already have a `rolename` property as well). maybe `find_delegation()` or something to show that we're looking for a delegation for the specific target_filepath?

I mention this elsewhere but this should probably return `Optional[str]` since it may be be called for a normal delegation and I think returning a name if the delegation does not match feels wrong -- this would also mean is_delegated_path() could just be deprecated as this method returns the answer to that as well.

jku commented on tuf/api/metadata.py:
```diff
@@ -1289,32 +1291,59 @@ class DelegatedRole(Role):
     ):
         super().__init__(keyids, threshold, unrecognized_fields)
+        if name is None and succinct_hash_info is None:
+            raise ValueError(
+                "Name must be set when succinct_hash_info is not used"
+            )
         self.name = name
         self.terminating = terminating
         if paths is not None and path_hash_prefixes is not None:
             raise ValueError("Either paths or path_hash_prefixes can be set")
+        if paths is not None and succinct_hash_info is not None:
+            raise ValueError("Either paths or succinct_hash_info can be set")
+        if path_hash_prefixes is not None and succinct_hash_info is not None:
+            raise ValueError(
+                "Either path_hash_prefixes or succinct_hash_info allowed"
+            )
 
-        if paths is None and path_hash_prefixes is None:
-            raise ValueError("One of paths or path_hash_prefixes must be set")
+        if (
+            paths is None
+            and path_hash_prefixes is None
+            and succinct_hash_info is None
+        ):
+            raise ValueError(
+                "paths, path_hash_prefixes or succinct_hash_info required"
+            )
```
if you want to remove some code, something like this could work:
```python
exclusive_vars = [paths, path_hash_prefixes, succinct_hash_info]
if sum(1 for var in exclusive_vars if var is not None) != 1:
    raise ValueError("One of (paths, path_hash_prefixes, succinct_hash_info) must be set")
```

MVrachev replied:
Loved this simplification.
Great suggestion!

jku reviewed (changes requested):
Nice work. 
* I left a complaint about using dict to "hide" API, that one I fee strongly about, let's figure out a solution for it. 
* changing API in incompatible ways (like arg order in `DelegatedRole.__init__()`) may be inevitable, but let's at least try to avoid it
* other comments are not strictly change requests.

Is the current implementation using None as the key in `Delegations.roles` dictionary? I'm quite surprised mypy does not complain about it in Delegations.from_dict() ...

Commit:
Add SuccinctHashDelegations class

Add SuccinctHashDelegations class containing the information from the
succinct_hash_delegation dict described in TAP 15.
This allows for easy mypy checks on the types, easy enforcement on
TAP 15 restrictions (as for example that "delegation_hash_prefix_len"
must be between 1 and 32) and support for unrecognized fields
inside succinct_hash_delegation without much of a hassle.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have made multiple changes:
1. Made `DelegatedRole.name` as the first attribute in the initialization list (the same way as it's now in the develop branch) and as thus removing the incompatible change
2. Renamed `DelegatedRole.get_rolename` to `Delegated.find_delegations` and used  `find_delegations` replacing the need for `DelegatedRole.is_delegated_path`.
3. Simplified checks with `sum` suggestions by @jku 
4. Experimented with creating a new container helper class `HashDelegations` containing the information about succinct hash delegations. This allowed me to make sure the requirements from TAP 15 are met inside the class without making `DelegatedRole` even more complicated and to add support for unrecognized fields inside the `succinct_hash_delegations` dict.

@jku it's ready for another review.

jku commented on tuf/api/metadata.py:
```diff
@@ -1313,8 +1380,16 @@ def __init__(
+            if not isinstance(succinct_hash_info, Dict):
+                raise ValueError("succinct_hash_info must be a dict")
+
+            succinct_info = HashDelegations.from_dict(succinct_hash_info)
```
you're calling `from_dict()` from a constructor, this is not how any existing code does it. from_dict() is a helper for the deserialization path but constructors can be called elsewhere too: constructors don't take unprocessed json dictionaries as argument, they take deserialized values/objects (like your new HashDelegations) as arguments.

jku commented on tuf/api/metadata.py:
```diff
@@ -1384,6 +1465,58 @@ def _is_target_in_pathpattern(targetpath: str, pathpattern: str) -> bool:
+
+        return None
+
     def is_delegated_path(self, target_filepath: str) -> bool:
```
So with this design:
* is_delegated_path() could be deprecated: find_delegation() answers that question and supports succinct delegations
* however is_delegated_path() is still available and currently returns wrong answer if used with a succinct delegation

I think we should document is_delegated_path as deprecated in the docstring and maybe add the couple of lines needed to fix the second issue as well

jku commented on tuf/api/metadata.py:
```diff
@@ -1384,6 +1465,58 @@ def _is_target_in_pathpattern(targetpath: str, pathpattern: str) -> bool:
+        If the given ``target_filepath`` is not in any of the paths that
+        ``DelegatedRole`` is trusted to provide it returns None.
+
+        If succinct_hash_info is used the result will be in the format
+        "{bin_name_prefix}-{bin_number}" where "bin_name_prefix" is information
+        provided by succinct_hash_info and "bin_number" is the corresponding
+        bin number calculated based on the ``target_filepath``.
```
Nit: Looks complex and too detailed for this docstring, maybe not needed at all?

MVrachev replied:
Maybe you are right. Will remove it.

jku commented on tuf/api/metadata.py:
```diff
@@ -1384,6 +1465,58 @@ def _is_target_in_pathpattern(targetpath: str, pathpattern: str) -> bool:
 
         return True
 
+    def _calculate_delegation(self, hash_bits_representation: str) -> str:
```
this method now seems like it should be part of the new class (HashDelegations.find_delegation() maybe)?

I also think the argument to the function should be `targetpath` : this way the arguments and output make sense to a reader: you feed in the targetpath and get the real delegated role name as output -- The fact that this specific implementation uses a string of ones and zeros as an intermediate value is not important and you only need to move a few lines of code to get this better abstraction.

Then `DelegatedRole.find_delegation()` ends up being quite clean, something like
```
if self.succinct_hash_info:
    return self.succinct_hash_info.find_delegation(target_filepath)
elif self.is_delegated_path(target_filepath):
    return self.name
return None
```

Opinions?

jku replied:
To make this a little more complex: we don't have an API for someone who
* uses non-succinct hash delegations
* wants to find out which bin might provide (or shoudl provide) a targetpath

This API clearly needs to be above the DelegatedRole level -- at Delegations maybe, but internally could use some of the same code

This question is not really TAP 15 related... but makes sense to solve (or at least discuss) at same time (and is related to #1952 and warehouse work)

MVrachev replied:
The reason I don't pass `targetpath` as an argument is the testing.
I want to test the actual bin calculation math and if I pass the `targetpath` this means that the hash calculation has to be moved inside `_calculate_delegation` and I won't be able to pass the binary representation in a string as do now. 
That way the test inside `test_api.py` won't be clear for the reader if we do that.
Given that this is an internal API I think it's allowed to require a little more specific argument for easier testing.

jku replied:
hmm, I think I'll leave this to you and @lukpueh as I think you've been talking about this part of the code.

Personally I'm not fond of making production code harder to read based on unit testing needs, or the basic idea of building a string of ones and zeros just to turn it into an int later ... but I can live with those so I'll leave these for you

jku commented on tuf/api/metadata.py:
```diff
@@ -1258,6 +1258,64 @@ def to_dict(self) -> Dict[str, Any]:
         return snapshot_dict
 
 
+class HashDelegations:
+    """A container with information about a succinct hash delegation.
```
I think I like this: I imagined it would be OO-overkill but maybe not: 

We could move some functionality here 
* like the delegation math you need, see other comment: I think it'll look just fine
* but also the math required to create new succinct delegations -- where @kairoaraujo can say something like _"I want 2**14 bins, give me the delegation but also give me a generator for all the bin names so I can create the metadata for those"._

We'll want to be careful with the name though, there are so many delegation related names. We now have three completely different names for the same concept, this does not feel correct: 
* spec calls it `succinct_hash_delegations`
* DelegatedRole calls it `succinct_hash_info`
* here it's called `HashDelegations`

MVrachev replied:
I agree about the naming.
Honestly I wasn't sure how to call it at all.
Maybe `SuccinctHashInfo`? But I don't like the suffix `info` as it doesn't bring any new context...
On another hand, `SuccinctHashDelegations` becomes too much...

jku replied:
On the other hand, class name is not used often so that could still work if you can keep the variable name as something clearly related but a little shorter, something like:

succinct_delegations = SuccinctHashDelegations(3, "bins")

Admittedly that is a bit longer than your current variable name

jku commented on tuf/api/metadata.py:
```diff
@@ -1274,12 +1332,15 @@ class DelegatedRole(Role):
 
     Args:
-        name: Delegated role name.
+        name: Delegated role name. If succinct_hash_info is passed, then a None
+            value should be given to ``name``.
```
Note that there is quite a bit of text on the mutual exclusivity of arguments above this: probably makes sens to mention this name issue there and fix that text in general.

jku commented on tuf/api/metadata.py:
```diff
@@ -1289,22 +1350,28 @@ class DelegatedRole(Role):
         terminating: bool,
         paths: Optional[List[str]] = None,
         path_hash_prefixes: Optional[List[str]] = None,
+        succinct_hash_info: Optional[Dict[str, Any]] = None,
         unrecognized_fields: Optional[Mapping[str, Any]] = None,
```
strictly speaking this is still breaking API.  
`DelegatedRole("a", [], 1, False, [], None, {})`
used to be valid  code. Now it will fail I assume. 

New arguments should generally speaking be added in the end of the list... I'm not saying we can't break API here: it's not a major breakage by any means but maybe I'd still vote for adding the new arg in the end.

MVrachev replied:
I didn't add it to the end as in my mind `unrecognized_fields` are at the end of each class, but maybe that's not so important.

jku reviewed (commented):
Thanks! I think the direction in this version can work. It's a new class just to hold two variables: this feels like overkill at first but there's some specific problems (the bin math plus the handling unrecognized fields in the new dict) that we can contain within that class so I think it works out.

Commit:
Deprecate DelegatedRole.is_delegated_path()

Deprecate DelegatedRole.is_delegated_path() in favor of
DelegatedRole.find_delegations() which will not only check if the
given targetpath is one of the paths DelegatedRole is trusted to
provide, but it will also return the role name.
This is especially useful when using succinct hash delegations.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
@jku I updated the pr once again making the following changes (not in order):
- renaming `HashDelegations` to `SuccinctHashDelegations` - please leave a comment what do you think
- left a deprecation messaged for `is_delegated_path` and made `is_delegated_path` work when `succinct_hash_delegations` is used
- removed a complex comment inside `from_delegations`
- moved the helper function `_calculate_delegation` inside `SuccinctHashDelegations`
- added additional comments about the exclusive arguments in the `DelegatedRole` docstring

MVrachev commented:
Made an update to one of the commits to remove the `f` string feature that caused problems for all `python3.7` CI runs.

jku commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,99 @@ def to_dict(self) -> Dict[str, Any]:
 
 
+class SuccinctHashDelegations:
+    """A container with information about a succinct hash delegation.
```
I think this still needs an explanation: "succinct hash delegation" is probably word salad for anyone except about 20 people on the planet.

It does not need to explain every detail but the core idea: that it desribes delegations of all possible target paths to variable number of bins, the names of which are generated based on the two fields.

jku commented on tests/test_metadata_eq_.py:
```diff
@@ -182,6 +183,23 @@ def test_snapshot_eq_(self) -> None:
         self.assertNotEqual(md.signed, signed_copy)
 
+    def test_hash_delegations_eq(self) -> None:
+        succinct_hash_info_dict = {
+            "delegation_hash_prefix_len": 8,
+            "bin_name_prefix": "delegated",
+        }
+        hash_deleg = SuccinctHashDelegations.from_dict(succinct_hash_info_dict)
```
not terribly important but I think it's weird to use `from_dict()` outside of serialization testing: from_dict() is a serialization implementation detail, and using the constructor should be less lines of code as well.

This style is used in multiple tests

jku commented on tuf/api/metadata.py:
```diff
@@ -1326,21 +1418,31 @@ def __eq__(self, other: Any) -> bool:
+        succinct_hash_dict = role_dict.pop("succinct_hash_delegations", None)
+        succinct_hash_info = None
+        if succinct_hash_dict is not None:
+            if not isinstance(succinct_hash_dict, Dict):
+                raise ValueError("succinct_hash_delegations must be a dict")
```
maybe not needed: the constructor will fail if it's not a dict

MVrachev replied:
I added this as the error I got was an `AttributeError: 'str' object has no attribute 'pop'` ` and I thought that it made sense to give a more concrete message, but python developers will know what that message means, so it probably it doesn't make sense.
Will remove this check.

jku commented on tuf/api/metadata.py:
```diff
@@ -1384,8 +1490,39 @@ def _is_target_in_pathpattern(targetpath: str, pathpattern: str) -> bool:
+
     def is_delegated_path(self, target_filepath: str) -> bool:
-        """Determines whether the given ``target_filepath`` is in one of
+        """This method is depricated, use find_delegation instead!
```
```suggestion
        """This method is deprecated, use ``find_delegation()`` instead.
```

fixed the typo, but while I was doing that also removed the exclamation: no need to shout in docs. Same typo in a warning message.

There might be some fancy way to product nice note/warning layout in RTD... but I think this is fine, no need to overdo it

jku commented on tuf/api/metadata.py:
```diff
@@ -1399,6 +1536,9 @@ def is_delegated_path(self, target_filepath: str) -> bool:
                 targets URL.
         """
 
+        msg = "is_delegated_path is depricated, use find_delegation instead!"
+        warnings.warn(msg, DeprecationWarning)
```
it's not quite this simple if you want to keep using this method in find_delegation() for now: now calling find_delegation(), a non-deprecated function, will result in a deprecation warning.

So either we don't have the warning (this is fine IMO: switching from `is_delegated_path()` to `find_delegation()` does not require any preparation) or we prevent the warning in the non-deprecated code path.

jku reviewed (commented):
* SuccinctHashDelegations works better as a name I think: it's long but not used v commonly.
* moving `_calculate_delegation()` without modifying it otherwise probably does not make sense:  having to disable pylints is a good hint that something is not right. my opinion is still that it would make sense for SuccinctHashDelegations to expose an actual delegation lookup method that accepts target path as argument but I'll just leave that as opinion
* I think it mostly looks good: our naming is now not perfect as DelegatedRole can now represent multiple roles and as result we need `DelegatedRole.find_delegation()`  which really does not make sense as a name ... but it seems best we can do without breaking API: I'm fine with that solution

Commit:
Add SuccinctHashDelegations.find_bin()

Add a public method to calculate and receive the bin name when succinct
hash delegation is used.
It's logical that all of the logic required to calculate the bin name
should be inside the class responsible for that SuccinctHashDelegations.

This will eliminate the need of
SuccinctHashDelegation._find_bin_for_bits() and its test.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add SuccinctHashDelegations.get_bin_name()

Add a SuccinctHashDelegations.get_bin_name() API call which returns the
bin name of a given sequential bin number.
This is useful to easily get the name of the first bin which can be
used as a unique identifier for delegated roles using a succinct hash
delegations.

A unique identifier for such delegated roles with succinct hash
is required by our implementation of Delegations.roles where the keys
are DelegatedRole.name attributes and values are DelegatedRole objects
and the fact that when succinct hash delegations are used
DelegatedRole.name should be populated with "None".

Solves: https://github.com/theupdateframework/python-tuf/issues/1943

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I addressed all of the @jku comments.
I found a way to add a good looking warning by using the `sphinx` format: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#warnings which looks like this: 
![image](https://user-images.githubusercontent.com/16246778/164280170-f2d8f9e3-c172-469d-9948-7a3773bf9528.png)
Additionally, I added the proposed solution by @jku inside #1943 to use the first bin name as the unique identifier inside `DelegatedRole.name` when succinct hash delegation is used.

@lukpueh and @jku it's ready for a review.

jku commented on tuf/api/metadata.py:
```diff
@@ -1384,8 +1510,33 @@ def _is_target_in_pathpattern(targetpath: str, pathpattern: str) -> bool:
+        """
+        # Check if the given target_filepath is in any of the "paths" or
+        # "path_hash_prefixes" that DelegatedRole is trusted to provide.
+        warnings.filterwarnings("ignore", category=DeprecationWarning)
```
It's still not this simple. This filters all deprecation warnings within the application from this point forwards, this is incorrect for two reasons:
* we want to filter only warnings from this module
* we want to filter for only a single call, not all calls from now on

You would need to manually remove the filter, or better yet use the catchwarnings() context manager. 

This complexity is why I'm not necessarily convinced the deprecation warnings are worth it: they are surprisingly difficult to get right.

MVrachev replied:
Okay, for now, I will keep it simple and leave a docstring warning.
If we later decide to add a deprecation warning it could be worth it to explore, but probably for a more critical API than an API from a helper class.

jku commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+            **self.unrecognized_fields,
+        }
+
+    def get_first_bin_name(self) -> str:
+        """Return the name of the first bin"""
+        return f"{self.bin_name_prefix}-0"
```
This does not seem to need a method *if* it's this simple...

However, this result (and _calculate_delegation() return value) seemto not match TAP example or the previous implementation IIRC: I'm leaving a longer comment in _calculate_delegation

lukpueh replied:
I agree with @jku that this might not need a dedicated method. I think a `get_bin_name(self, bin_number)` method would be more useful, especially given the subtleties (zero-padded hex) about the name that Jussi mentioned.

MVrachev replied:
Agreed that `get_bin_name(self, bin_number)` sounds a lot better.
Will change this method.

jku commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+
+        bit_length = self.hash_prefix_len
+        # Get the first bit_length of bits and then cast them to decimal.
+        bin_number = int(hash_bits_representation[:bit_length], 2)
+
+        name_prefix = self.bin_name_prefix
+        return f"{name_prefix}-{bin_number}"
```
This does not seem to be in line with the TAP examples and if I recall correctly the previous implementation: 

TAP definition:
> The name will be structured as NAME_PREFIX-COUNT where COUNT is a hexadecimal value between 0 and 2^BIT_LENGTH-1 (inclusive) that represents the bin number

TAP example:
>      "name": alice.hbd-00,

 
Your rolename seems to be decimal and not padded with zeros. Is this correct?

lukpueh replied:
Minor style nit: Is it really necessary to alias these names here? Wouldn't it be okay to:
```python
# Cast relevant bit prefix to decimal to get the bin number
bin_number = int(hash_bits_representation[:self.hash_prefix_len], 2)
return f"{self.bin_name_prefix}-{bin_number}" # FIXME: Format as zero-padded hex
```

jku reviewed (commented):
I have not looked at all of it but the naming scheme seems important enough to raise the issue now (see code comments)

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+
+    Args:
+        hash_prefix_len: Number of bits between 1 and 32 used to separate the
+            different bins.
```
I think this needs a better explanation, either here or in the function description above.

MVrachev replied:
Better explanation about what is a hash deletion or specifically about succinct hash bin delegation?
Can you give some reference points to what you mean?

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+        if hash_prefix_len <= 0 or hash_prefix_len > 32:
+            raise ValueError("hash_prefix_len must be between 1 and 32")
+        if not isinstance(bin_name_prefix, str):
+            raise ValueError("bin_name_prefix must be a string")
```
May I ask why you don't check if `hash_prefix_len` is an int, but do check if `bin_name_prefix` is a string?

MVrachev replied:
If `hash_prefix_len` is not a number a comparison like the one done on line 1283 will throw a `TypeError`, for example if `hash_prefix_len` is a string this comparison will throw `TypeError: '<' not supported between instances of 'str' and 'int'`
I am not convinced there is a point in checking that `hash_prefix_len` is an `int`, not a `float`.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+
+        self.hash_prefix_len = hash_prefix_len
+        self.bin_name_prefix = bin_name_prefix
+        self.unrecognized_fields: Mapping[str, Any] = unrecognized_fields or {}
```
See https://github.com/theupdateframework/python-tuf/pull/1963

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+        hasher = sslib_hash.digest(algorithm="sha256")
+        hasher.update(target_filepath.encode("utf-8"))
+        # Get the bit representation of the hash.
+        hash_bits = "".join(f"{one_byte:08b}" for one_byte in hasher.digest())
```
Minor optimization nit: You only need the first `math.ceil(self.hash_prefix_len / 8)` bytes.

MVrachev replied:
I am not sure I understand your point here.
If `hash_prefix_len = 17` and I do that math it means `hash_bits` will be `3`.
Meaning I won't be comparing the bits with all of the bits I have defined in `succinct_hash_delegations["delegation_hash_prefix_len"]`.

Maybe I can look only into the `hash_prefix_len` amount of bytes without dividing by 8?

lukpueh replied:
> If hash_prefix_len = 17 and I do that math it means hash_bits will be 3.

No, it means that you only need the first 3 bytes in order to look at the first 17 bits (all bits from first and second byte + the left-most bit from the third byte).

> Maybe I can look only into the hash_prefix_len amount of bytes without dividing by 8?

You don't need 17 bytes to get the left-most 17 bits.

This should give you the desired prefix of the hash:

```python
needed_bits = self.hash_prefix_len # aliased for emphasis
needed_bytes =  math.ceil(needed_bits / 8)

prefix = "".join(f"{one:08b}" for one in hasher.digest()[:needed_bytes])[:needed_bits] # oneliner might be a bit ugly
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+        # Get the bit representation of the hash.
+        hash_bits = "".join(f"{one_byte:08b}" for one_byte in hasher.digest())
+
+        # Split hash calculation and from the rest for easier testing.
```
```suggestion
        # Split hash calculation from the rest for easier testing.
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+        name_prefix = self.bin_name_prefix
+        return f"{name_prefix}-{bin_number}"
+
+    def find_delegation_bin(self, target_filepath: str) -> str:
```
Minor naming nit: Maybe `find_bin` is enough?

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1259,122 @@ def to_dict(self) -> Dict[str, Any]:
+        """Return the name of the first bin"""
+        return f"{self.bin_name_prefix}-0"
+
+    def _calculate_delegation(self, hash_bits_representation: str) -> str:
```
Minor naming nit: `_calculate_delegation` sounds a little vague. Maybe `_find_bin_for_bits` makes it clearer that this is part of `find_bin` below?  🤷

lukpueh reviewed (commented):
Awesome work! I love how all of this can be implemented in the metadata classes and does not need any changes in `ngclient`.

I mostly reviewed the `SuccinctHashDelegations` class in this round and left a couple of minor comments, let me know what you think.

Regarding the changes in `Delegations` and `DelegatedRole`, I lean towards revisiting the TAP 15 metadata definition. See my suggestion for an alternative format that might make the implementation a little less "hackish" in: https://github.com/theupdateframework/python-tuf/issues/1943#issuecomment-1106390045

I haven't looked closely at the tests yet, but will follow up shortly. Do you plan to add a client test that validates a target delegated to a hash bin?

Commit:
SuccinctHashDelegations: add zero padding to bins

Add zero padding to bin names inside SuccinctHashDelegations.
Zero padding ensures that the bin names always have the same length.

This characteristic is implied in the example given by TAP 15 where
the third bin is named "alice.hbd-03"
For context read TAP 15:
https://github.com/theupdateframework/taps/blob/master/tap15.md

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add helper methods in SuccinctHashDelegations

Add get_all_bin_names and is_bin in SuccinctHashDelegations.
Those methods proved useful in the testing code, but I believe they have
a potential value for production code as well.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
verify_delegate() with succinct hash delegation

Change Metadata.verify_delegate(), so it can handle delegated roles
where succinct hash delegation is used.
In the case of succinct hash delegation, when the delegator is "targets"
we have to make sure that a given role "r" contains a succinct hash
delegation information and that the "delegated_role" is a role name is a
bin handled by "r".
This complication comes from the idea that we could have two delegated
roles with succinct hash delegation information and there could be
other non-succinct roles as well.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Simulator to work with succinct hash delegations

Change RepositorySimulator.add_delegations() to work with delegated
role "role" that uses succinct hash delegations.
This case is special as for that role we are not creating one target,
but we are creating target metadata instances for all bins.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Test delegation tree with succinct hash delegation

Test traversing the delegation tree when there is a delegated role
using succinct hash delegation.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I have updated my pr and addressed all comments made by @jku and @lukpueh.
I am preparing another commit following @lukpueh suggestions in https://github.com/theupdateframework/python-tuf/issues/1943#issuecomment-1106390045 and I am will commit it in another branch.

jku commented on tests/repository_simulator.py:
```diff
@@ -346,9 +346,16 @@ def add_target(self, role: str, data: bytes, path: str) -> None:
-        """Add delegated target role to the repository."""
+        """Add delegated target role to the repository.
+
+        If no "targets" are provided this indicates "role" uses a succinct hash
+        delegation and target metadata instances will be created for all bins.
```
this is untrue according to code: whether targets are provided is unrelated to succinct-or-not question

jku commented on tests/test_updater_delegation_graphs.py:
```diff
@@ -425,6 +426,94 @@ def test_hash_bins_graph_traversal(
+            target_path="missingpath",
+            expected_visited_bin="delegated-2",
+        ),
+        "case 6": SuccinctHashTestCase(
+            hash_prefix_len=2,
+            target_path="bar",
+            expected_visited_bin="delegated-3",
+        ),
```
nice test! Have you tried if it's slow with "real" numbers like if your length is 8 or 16?

i am asking because these artificially small test cases mean reader can't easily see if the number is hexadecimal or decimal (and the padding does not get tested)...

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1267,173 @@ def to_dict(self) -> Dict[str, Any]:
+        # If hash_prefix_len = 8 then number_of_bins = 256 and zero_padding = 3
+        # meaning the third bin will have a suffix of "003".
+        self.number_of_bins = 2**hash_prefix_len
+        self.zero_padding = int(math.log10(self.number_of_bins)) + 1
```
Ha that's smart!  Should give the same result as `len(str(self.number_of_bins))`, right? I wonder if using `len` doesn't make it slightly easier to understand (and might also be more performant). But I'm fine either way.

Minor naming nit: Maybe something like `suffix_width` or so makes more sense? Because the amount of needed zero padding actually is `(int(math.log10(self.number_of_bins)) + 1) - len(bin_index)`.

jku replied:
it's still decimal though and not hex...

I'm sorry if I sound like a broken record but I'm not sure if this has been addressed anywhere.  Maybe I'm missing something but I'd like to see this addressed in a comment at least: the spec seems to define the bin names containing hexadecimal yet the PR uses decimal?

lukpueh replied:
You're right, TAP 15 says "*NAME_PREFIX-COUNT where COUNT is a **hexadecimal** value between 0 and 2^BIT_LENGTH-1*". So my suggestion would be `len(f"{self.number_of_bins:x}"))`.

MVrachev replied:
Oh...totally missed that...

MVrachev replied:
I realized that besides the calculation of the zero-padding I will have to change the calculation of `suffix` in `get_bin_name` API call in order to save the suffix as HEX.

jku replied:
I assume the reason for using hex here is that it makes the binary math a little easier for humans to verify: if your targetpath hash begins with "a1d8...", then in a 16 bit succinct delegation the correct bin name should be "prefix-a1d8" -- at least that's how I've understood it

jku commented on tuf/api/metadata.py:
```diff
@@ -421,6 +421,13 @@ def verify_delegate(
 
             keys = self.signed.delegations.keys
             role = self.signed.delegations.roles.get(delegated_role)
+            if role is None:
+                for r in self.signed.delegations.roles.values():
+                    if (
+                        r.succinct_hash_info is not None
+                        and r.succinct_hash_info.is_bin(delegated_role)
+                    ):
+                        role = r
```
I think this makes a good point for possibly simplifying the TAP: this is probably close to the best we can do with the current design but it really isn't obvious why we do this in the order we do and why we loop through all delegated roles

(the alternative implementation that avoids looping through all roles for the current TAP design would take `delegated_role`, try to parse it as a succinct delegation, translate "prefix-NNN" to "prefix-0", try to lookup "prefix-0" in `delegations.roles`, etc: still not simple or obvious)

jku replied:
and if it's not obvious to the reader: the verifying code needs a way to find a succinct delegation based on a rolename, assuming the succinct delegation delegates to that role name... this is not trivial when there can be other, normal, delegations and multiple succinct delegations in the same Delegations structure.

MVrachev replied:
I will leave a comment to be a little clearer.

jku commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1267,173 @@ def to_dict(self) -> Dict[str, Any]:
+        hasher.update(target_filepath.encode("utf-8"))
+        # Get the bit representation of the hash. The bit representation is
+        # calculated for a small number of symbols required by hash_prefix_len.
+        needed_bits = self.hash_prefix_len  # aliased for emphasis
+        needed_bytes = math.ceil(needed_bits / 8)
+        hash_bits = "".join(
+            f"{one_byte:08b}" for one_byte in hasher.digest()[:needed_bytes]
+        )
+        needed_hash_bits = hash_bits[:needed_bits]
+
+        # Split hash calculation from the rest for easier testing.
+        return self._find_bin_for_bits(needed_hash_bits)
```
I've not played with this yet but after looking at this and talking to Lukas, I think I finally understood what's being done here: it's essentially a right shift to get hash_prefix_len leftmost bits. I think if this is the case then I would prefer seeing the actual math here without the string-of-1-and-0.

```python
# we can't ever need more than 4 bytes (32 bits)
hash_bytes = hasher.digest()[:4]

# right shift so that we only have the leftmost hash_prefix_len bits we care about
bin_number = int.from_bytes(hash_bytes, byteorder="big") >> (32 - self.hash_prefix_len)
        
return self.get_bin_name(bin_number)
```

**EDIT** I tested and that does work (after I fixed the confusion I was having about digest() return value)

lukpueh replied:
🤦  "Obviously"  bit shifting is the way to do this.

MVrachev replied:
You are correct that directly shifting the bits is easier to understand.
Will apply this. Thanks for the detailed comment!

jku commented on tuf/api/metadata.py:
```diff
@@ -1258,17 +1267,173 @@ def to_dict(self) -> Dict[str, Any]:
+        needed_bits = self.hash_prefix_len  # aliased for emphasis
+        needed_bytes = math.ceil(needed_bits / 8)
+        hash_bits = "".join(
+            f"{one_byte:08b}" for one_byte in hasher.digest()[:needed_bytes]
```
i'm not sure if this is a bug or just bad naming but a hexadecimal digit is not equal to a byte (but see my other comment: I'm proposing removing this code altogether)

jku replied:
oh I may be misunderstanding the hashlib use here -- digest() maybe actually returns the bytes and not hex: in that case never mind me, I'm the one making a mistake

EDIT: Yes, my bad: I'm marking this resolved

Commit:
Test find_delegation from target path to bin name

Change the test_find_delegation_with_succinct_hash_info() test, so it
can test the whole calculation of bin name from a given target path.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
SuccinctHashDelegations.find_bin() use bit shift

Remove the need of calculation of all hash bits and then filtering
those that aren't needed with simple bit shifting.
That way it's clearer what we actually do without using complicated
python tricks.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
I think I have applied all suggestions by @jku  and @lukpueh.
Most notably I changed:
- the bin names suffix is now hex
- calculating the bin names is done by bit shifting
- `test_find_delegation_with_succinct_hash_info` is changed, so it's testing the whole process of calling `find_delegations` to reciving the actual bin name
- added multiple bin names case inside `test_succinct_hash_bins_graph_traversal` to make sure all bin files are existing in cases where zero padding has to be used (the bins are more than 10).

MVrachev commented:
I have implemented @lukpueh suggestion from the comment here https://github.com/MVrachev/tuf/tree/new-suggestion.
For those interested, please comment on issue #1943.

MVrachev commented:
Together with @jku, @lukpueh, and the TAP author - @mnm678 we met to discuss if we should change the design on TAP 15 based on multiple suggestions.
We end up deciding that there are multiple changes that we want to make which are all listed in my comment [here](https://github.com/theupdateframework/taps/pull/154#pullrequestreview-973613582).
Marina updated the TAP 15 content in her pr https://github.com/theupdateframework/taps/pull/154 and there were multiple changes that I had to do in this pr to be up to date.
That's why I decided it's probably best if I close this pr, make the changes from scratch (reusing big chunks of this code of course) and present a cleaner history in a new pr.
This pr is superseded by https://github.com/theupdateframework/python-tuf/pull/2010
